### ISSUE TITLE
Display whats now playing in music channel topic

### ISSUE THREAD

**Issue by Macley-Kun** [Frederikam/FredBoat#300]
Something like the picture down below.

![schermafbeelding 2017-09-18 om 17 30 12](https://user-images.githubusercontent.com/26381427/30550196-0ceeda44-9c97-11e7-8da2-cc036bfc8010.png)

I think it looks great, you can immediately see what's playing and by who!

**Comment by Macley-Kun** [comment]
Also congrats on your #300'd issue 🎉 🎉 🎉

**Comment by napstr** [comment]
Can't do this due to ratelimits last I checked.

**Comment by Macley-Kun** [comment]
Rip then nothing we can do about it :) Kinda strange however b/c another project i'm familiar with pulls this off 🤔 

Mind if i link that project or should i just close and forget about it :') ?

**Comment by napstr** [comment]
If you want to advance this issue you need to dig up the ratelimits and make an estimate whether it's doable on the public bot.

**Comment by Macley-Kun** [comment]
Mhm sure, i can ask the other guy about the in's and outs about that. Think it's worth the time :)

**Comment by Macley-Kun** [comment]
So i did some searching but can't rlly find a number, can you give me the link your used to check it Napstr?

**Comment by Frederikam** [comment]
https://discordapp.com/developers/docs/topics/rate-limits#header-format

**Comment by Macley-Kun** [comment]
Mhm so i was looking at the right page... But you never know the number unless you look at the header in your program/code log.

I heard that generally you can send 5 messages per 5 seconds.

**Comment by Macley-Kun** [comment]
um, idk what else i can add to this... Pre-assuming it's 5 topic changes per 5 seconds.

**Comment by Macley-Kun** [comment]
Sures ^_^ maybe it have to be a selfhost/patreon only thing. Thanks for headsup Shredder~

**Comment by napstr** [comment]
> Pre-assuming it's 5 topic changes per 5 seconds.
What's scope is this rate limit?

What's the source of this info?

**Comment by Macley-Kun** [comment]
@napstr i asked a guy who also is making a musicbot (also in java) for Discord.

**Comment by napstr** [comment]
> What's the scope of this rate limit?

**Comment by Macley-Kun** [comment]
Idk, the guy said 5 per 5 seconds :P

**Comment by napstr** [comment]
So we're back to here:

> If you want to advance this issue you need to dig up the ratelimits and make an estimate whether it's doable on the public bot.

Without knowledge of the (global) ratelimit for this action it's not possible to make an estimate.

**Comment by Macley-Kun** [comment]
Well ik but, i can only ask other people about it. I can't rlly develop my own application and look into the headers what the numbers are. I'm just a simple guy who suggested a feature.

I can try to ask it to the Discord API server.

**Comment by napstr** [comment]
Please refrain from bumping this issue then.

**Comment by Macley-Kun** [comment]
Asked in API, this is your answer you need Napstr :)

global ratelimit should be 50 per second. though that applies to all of your requests and doesn't imply that you should make any particular request 50 times in one second

**Comment by Frederikam** [comment]
The global rate limit is irrelevant in this case

**Comment by napstr** [comment]
If there is no special ratelimit on changing the channel topic that's fine, makes it a viable alternative to announcing the track.

**Comment by Nanabell** [comment]
So I was playing around with this idea a bit and want to report my findings

Firstly, it wasn't actually hard to implement a basic system that would update the current topic with the requested information 
![image](https://user-images.githubusercontent.com/13600006/31108019-d1bcd1a4-a7f6-11e7-9a8d-ff3553c38a9d.png) *(Not this is just a test not final design)*
it would update perfectly once a new song started playing and overall the feature should be pretty stable as its checking everytime a song starts if we still have permission to update the channel topic.

After that first successful test, I thought about implementing status updates for the different states a player can be in. eg `STOPPED`, `PAUSED`, `PLAYING` and made topic formats for each state and hooked them into the events respectively. There appeared the first issue, with the way we `skip` a track. When a song is skipped the player actually stopped and started the next song, which would mean we update the Topic twice just one after the other. (Note: This might also just be my fault in setting up a flawed testing environment)

The second and larger issue I started to notice was that the whole idea doesn't seem to work with how Fredboat is designed. I was talking with Napster about the idea a few days ago and we both agreed that we wanted to keep this whole feature low profile, meaning that we don't keep any new reference to any channel we want to update the topic of. Which implicitly meant that we have to use the currently active text channel of the player we already save to announce tracks etc. But since we do not keep any reference to the channel topic or even the old channel topic that we had overridden we would not be able to restore the old topic. And if someone started using the bot in a different channel we would override the second channel topic 
![image](https://user-images.githubusercontent.com/13600006/31108300-b937e996-a7f8-11e7-8359-f71f3ba08d90.png)
which would result in 2 channels having "playing" topics. Some music bots can be "bound" to a text channel and will only accept commands in that channel going forward, as such updating the channel topic is no big deal for them but Fredboat can and will override the "active" channel if it accepts new commands in a different channel.
We could ofc implement this if we increase the size of the whole system and keep reference a "locked" channel  where we always will change the topic regardless of what music is played or keep reference to the old topic message and restore that after we are done playing but that would be a lot more work

***tl; dr***:
The idea of updating a channel topic does not fit with the structure of Fredboat

**Comment by majorcyto** [comment]
I would like to put in a vote for the ability to "Bound" the bot to a channel actually. I know you can basically achieve the same use via permissions, but this is such a more straightforward way on the user side.

**Comment by Macley-Kun** [comment]
First off, i rlly wanna thank you Nanabell for testing, making and summarizing the request i made.
I think the first small issue is no problem. Fred noted "The global rate limit is irrelevant in this case". What means it's not a problem. The skip functionality is fast enough that people wouldn't notice it.

I have bin re-reading your summarize alot and this big issue about the channel topic thing is ofcours not passable. I do think it's maybe an option for users to set a: 'music topic'.

So image that we can just play, stop, pause music everywhere in the server. But the 'music topic' will only happen in the channel were the command is issued to enable it.

Use case:
Music-Channel
user: ;;entopic
bot: changes topic in Music-Channel from now on
(user starts, stops, pauses in another channel)
bot: changes topic in Music-Channel

But even if this would be approved. Storing the old topic after the music is done playing isn't rlly importend in my eyes. I image the user has already a dedicated music channel and just enables the functionality. No topic is set then.

If however we do want to store this. I think we have two options i can think so fast about:
1. We simpley just put the music topic before the setted topic:
Topic now: Type ;;smch to see all the music commands!
user: ;;entopic
Topic now: now playing <song name> <etc> Type ;;smch to see all the music commands!
(Ofcours this isn't the most elegant solution)
2. We output the current topic as a backup in the channel itself. If the music is done the user can manually then set the topic back.
user: ;;entopic
bot: the old topic: <old topic string> is replaced by the song that's currently played.

Sorry for the long post. But i hope i kinda made sence.

**Comment by napstr** [comment]
Lets save the channel that should be updated in the guild config.
